### PR TITLE
Use the same domain name for the connections

### PR DESCRIPTION
**Pull request by khanova**
Use the same domain name for the connections. It helps with connections caching on v8 side. 

Pass endpont id inside the options for ws connections.

https://github.com/neondatabase/neon/issues/6290

### CODE REVIEW

File: shims/net/index.ts [hunk 1]
```diff
@@ -62,17 +62,21 @@ export interface SocketDefaults {
 }
 type GlobalOnlyDefaults = 'poolQueryViaFetch' | 'fetchEndpoint' | 'fetchConnectionCache' | 'fetchFunction';
 
+const transformHost = (host: string): string => {
+  return host.replace(/^[^.]*/, 'api');
+}
```
Comment (jawj) [inline shims/net/index.ts:67]: Here and where we set the endpoint option, should we first check that the host has a plausible structure? For example, that it has at least two dots in it, or maybe even that it includes an AWS region? I ask because we advertise the driver as capable of being used with your own DB back-end and proxy, and this may break backwards-compatibility.

File: dist/npm/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@neondatabase/serverless",
-  "version": "0.7.2",
+  "version": "0.7.3",
```
Comment (jawj) [inline dist/npm/package.json:3]: Possibly this should be a minor rather than a patch level change, since it could plausibly break things for people?

**Review by jawj — COMMENTED** [review]
Thanks Anna. Please see individual comments.

File: export/index.ts [hunk 3]
```diff
@@ -81,6 +82,15 @@ class NeonClient extends Client {
...
     ) throw new Error(`No database host or connection string was set, and key parameters have default values (host: localhost, user: ${defaultUser}, db: ${defaultUser}, password: null). Is an environment variable missing? Alternatively, if you intended to connect with these parameters, please set the host to 'localhost' explicitly.`);
 
+    if (this.host !== 'localhost') {
+      // @ts-ignore -- TS doesn't know about this.connectionParameters
+      if (typeof this.connectionParameters.options !== 'string') {
+        // @ts-ignore
+        this.connectionParameters.options = '';
+      }
+      // @ts-ignore
+      this.connectionParameters.options += 'project=' + this.host.split('.')[0];
```
[the remark this answers is not in the document]
Comment (khanova) [reply export/index.ts:92]: (1) Currently proxy uses the same logic for both: `project` and `endpoint`, but sure, let's use `endpoint` since the project is a legacy option.

(2) Good point, let me fix this.

File: dist/npm/package.json (hunk 2 above)
Comment (khanova) [reply dist/npm/package.json:3]: I think, that with your suggestions it shouldn't break anything or am I missing something?

File: shims/net/index.ts (hunk 1 above)
Comment (jawj) [reply shims/net/index.ts:72]: Thanks for making the change, but I don't know that this works. Some DBs are Vercel Postgres DBs, and will be accessed at e.g. `my-endpoint.eu-central-1.postgres.vercel-storage.com`. Presumbly similar things may happen in future. So maybe just checking for two dots (and calling this a possibly-breaking change) is the best we can do?

**Comment by kelvich** [comment]
@duskpoet can you take a look this too? extra pair of eyes would be helpful

File: DEVELOP.md [hunk 4]
```diff
@@ -11,7 +11,7 @@ npm version patch  # or minor or major
 npm publish
 ```
 
-* To run or deploy the simple test app on Cloudflare, create a `.dev.vars` file containing `DATABASE_URL=postgres://connection_string`, run `npx wrangler dev --local` or `npx wrangler publish`.
+* To run or deploy the simple test app on Cloudflare, create a `.dev.vars` file containing `NEON_DB_URL=postgres://connection_string`, run `npx wrangler dev --local` or `npx wrangler publish`.
```
Comment (duskpoet) [inline DEVELOP.md:14]: why this change? Is it intentional?

File: dist/npm/package.json (hunk 2 above)
Comment (duskpoet) [reply dist/npm/package.json:3]: It's a pretty significant change, but I suggest to keep it as patch, to ensure that all who use the package would get the update at least on package-lock regeneration, and by dependabot pick up

Comment (duskpoet) [reply dist/npm/package.json:3]: I also expect that there are no breaking changes in the pr (didn't look through all of it yet)

File: export/index.ts [hunk 5]
```diff
@@ -81,6 +82,22 @@ class NeonClient extends Client {
       this.password === null
     ) throw new Error(`No database host or connection string was set, and key parameters have default values (host: localhost, user: ${defaultUser}, db: ${defaultUser}, password: null). Is an environment variable missing? Alternatively, if you intended to connect with these parameters, please set the host to 'localhost' explicitly.`);
 
+    if (isNeonHost(this.host)) {
+      const endpointOption = 'endpoint=' + this.host.split('.')[0];
+      // @ts-ignore -- TS doesn't know about this.connectionParameters
+      if (typeof this.connectionParameters.options !== 'string') {
```
Comment (duskpoet) [inline export/index.ts:88]: I'd suggest to add it to the class declaration, it would be safer

File: shims/net/index.ts [hunk 6]
```diff
@@ -62,17 +62,26 @@ export interface SocketDefaults {
 }
 type GlobalOnlyDefaults = 'poolQueryViaFetch' | 'fetchEndpoint' | 'fetchConnectionCache' | 'fetchFunction';
 
+export const isNeonHost = (host: string) => {
+  return host.endsWith('.aws.neon.tech') || host.endsWith('.aws.neon.build');
```
Comment (duskpoet) [inline shims/net/index.ts:66]: this wont work for vercel databases, right? @kelvich

File: shims/net/index.ts (hunk 1 above)
Comment (duskpoet) [reply shims/net/index.ts:72]: Our endpoints have a distinctive form of ep-{word}-{word}-{alphanumerics}. Maybe we can rely on that?

File: dist/npm/package.json (hunk 2 above)
Comment (duskpoet) [reply dist/npm/package.json:3]: yea, we can break stuff. Let's do a minor version?

File: DEVELOP.md (hunk 4 above)
Comment (khanova) [reply DEVELOP.md:14]: Yes, the previous description was incorrect.

File: shims/net/index.ts (hunk 1 above)
Comment (khanova) [reply shims/net/index.ts:72]: Good idea!

File: dist/npm/package.json (hunk 2 above)
Comment (khanova) [reply dist/npm/package.json]: Done

File: DEVELOP.md (hunk 4 above)
Comment (duskpoet) [reply DEVELOP.md:14]: I assume, users can use any format they want, but ok

File: shims/net/index.ts [hunk 7]
```diff
@@ -61,18 +61,27 @@ export interface SocketDefaults {
   disableSNI: boolean;
 }
 type GlobalOnlyDefaults = 'poolQueryViaFetch' | 'fetchEndpoint' | 'fetchConnectionCache' | 'fetchFunction';
+const neonHostPattern = /^(?:ep-)?([a-z]+)-([a-z]+)-(\w+)/;
```
Comment (duskpoet) [inline shims/net/index.ts:64]: "ep-" part should be mandatory, no?

File: shims/net/index.ts (hunk 1 above)
Comment (jawj) [reply shims/net/index.ts:72]: Yes, as long as we're pretty sure that this format (a) won't change any time soon and (b) is true for all partners (including e.g Vercel Storage).

File: shims/net/index.ts (hunk 7 above)
Comment (khanova) [reply shims/net/index.ts:64]: No, if it's old project there is no endpoint.

File: DEVELOP.md (hunk 4 above)
Comment (khanova) [reply DEVELOP.md:14]: Unfortunately, no, I struggled a lot with using `DATABASE_URL` and it didn't work.

Comment (duskpoet) [reply DEVELOP.md:14]: oh wow, sorry, I was under impression that this is just an env to var to user in your backend

File: shims/net/index.ts (hunk 7 above)
Comment (duskpoet) [reply shims/net/index.ts:64]: Oh, the ones with project id as first part...
I am afraid that without ep part it's pretty general... wdyt @jawj ?

File: DEVELOP.md (hunk 4 above)
Comment (jawj) [reply DEVELOP.md:14]: Sorry @khanova, you were right to switch this. It _is_ just an arbitrary env var name, but it's one that's expected by the code in `src/index.ts`, for testing on Cloudflare.

**Review by jawj — APPROVED** [review]
This all looks sensible to me now.

In an ideal world we'd have a nice test suite to check against — and with @duskpoet's help we hopefully soon will!

@khanova I'm assuming (e.g. based on your trouble with `DATABASE_URL` vs `NEON_DB_URL`) that you've done some basic testing already? On that basis, I'm approving the PR.

File: shims/net/index.ts (hunk 7 above)
Comment (jawj) [reply shims/net/index.ts:64]: @duskpoet Yeah, false positives (which would screw things up for people using our driver with a non-Neon database) are certainly possible here.

But they should be very rare (because I don't think there are many of these users), and they're easily fixable by overriding the `fetchEndpoint` and/or `wsProxy` configuration options.

So I think this is probably the right balance? I guess we could also check for at least two more parts afterwards, i.e. `/^(?:ep-)?([a-z]+)-([a-z]+)-(\w+)[.][^.]+[.][^.]+/`, but I think this barely helps.

Comment (khanova) [reply shims/net/index.ts]: Applied the suggestion.

File: shims/net/index.ts (hunk 1 above)
Comment (duskpoet) [reply shims/net/index.ts:73]: yea, the endpoint id format is more or less stable in this sense.

**Review by duskpoet — APPROVED** [review]
